File: backend/app/services/scoring.py
PILLAR_WEIGHTS: dict[str, float] = {
    "transparency": 0.16,
    "reliability": 0.16,
    "security": 0.20,
    "privacy": 0.16,
    "safety_societal": 0.16,
    "excessive_agency": 0.16,
}

PILLAR_KEYS = list(PILLAR_WEIGHTS.keys())


def compute_composite(pillar_scores: dict[str, float]) -> float:
    return sum(pillar_scores[p] * w for p, w in PILLAR_WEIGHTS.items())


def validate_composite_tolerance(pillar_scores: dict[str, float], submitted: float) -> bool:
    computed = compute_composite(pillar_scores)
    if computed == 0:
        return submitted == 0
    return abs(computed - submitted) / computed <= 0.05


def get_grade(score: float) -> str:
    if score >= 900:
        return "A"
    if score >= 800:
        return "B"
    if score >= 700:
        return "C"
    if score >= 600:
        return "D"
    return "F"


def aggregate_pillar_scores(scores: list[dict[str, float]]) -> dict[str, float]:
    """Arithmetic mean per pillar across a list of scanner score dicts."""
    if not scores:
        return {}
    return {p: sum(s[p] for s in scores) / len(scores) for p in PILLAR_KEYS}

File: frontend/src/lib/scoring.ts
import type { Grade, PillarBreakdown } from "@/lib/types";

export const PILLAR_WEIGHTS: Record<keyof PillarBreakdown, number> = {
  transparency: 0.16,
  reliability: 0.16,
  security: 0.20,
  privacy: 0.16,
  safety_societal: 0.16,
  excessive_agency: 0.16,
};

export const PILLAR_LABELS: Record<keyof PillarBreakdown, string> = {
  transparency: "Transparency",
  reliability: "Reliability",
  security: "Security",
  privacy: "Privacy",
  safety_societal: "Safety & Societal Impacts",
  excessive_agency: "Excessive Agency",
};

export const PILLAR_KEYS = Object.keys(PILLAR_WEIGHTS) as (keyof PillarBreakdown)[];

export function getGrade(score: number): Grade {
  if (score >= 900) return "A";
  if (score >= 800) return "B";
  if (score >= 700) return "C";
  if (score >= 600) return "D";
  return "F";
}

export function isStale(scoredAt: string | null, thresholdDays = 90): boolean {
  if (!scoredAt) return false;
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - thresholdDays);
  return new Date(scoredAt) < cutoff;
}

export function hasDivergence(
  scanner: PillarBreakdown,
  aggregate: PillarBreakdown
): boolean {
  return PILLAR_KEYS.some((k) => Math.abs(scanner[k] - aggregate[k]) >= 100);
}

export function getDivergingPillars(
  scanner: PillarBreakdown,
  aggregate: PillarBreakdown
): (keyof PillarBreakdown)[] {
  return PILLAR_KEYS.filter((k) => Math.abs(scanner[k] - aggregate[k]) >= 100);
}
